ipc/src/win.rs

```diff
@@ -85,38 +85,47 @@ impl<M: Metadata, S: Middleware<M>> PipeHandler<M, S> {
 					.out_buffer_size(MAX_REQUEST_LEN)
 					.in_buffer_size(MAX_REQUEST_LEN)
 					.create()
-			),
+			)),
 			io_handler: io_handler.clone(),
 		})
 	}
 
-	fn handle_incoming(&mut self, addr: &str, stop: Arc<AtomicBool>) -> io::Result<()> {
+	pub fn stop(&self) {
+		self.waiting_pipe.lock().unwrap().disconnect().ok();
+	}
+
+	fn handle_incoming(&self, addr: &str, stop: Arc<AtomicBool>) -> io::Result<()> {
 		trace!(target: "ipc", "Waiting for client: [{}]", addr);
 
-		// blocking wait with small timeouts
-		// allows check if the server is actually stopped to quit gracefully
-		// (`connect` does not allow that, it will block indefinitely)
-		loop {
-			if let Ok(_) = NamedPipe::wait(addr, Some(200)) {
-				try!(self.waiting_pipe.connect());
-				trace!(target: "ipc", "Received connection to address [{}]", addr);
-				break;
-			}
-			if stop.load(Ordering::Relaxed) {
-				trace!(target: "ipc", "Stopped listening sequence [{}]", addr);
-				return Ok(())
-			}
-		}
+		let mut overlapped = ::miow::Overlapped::zero();
 
-		let mut connected_pipe = std::mem::replace::<NamedPipe>(&mut self.waiting_pipe,
-			try!(NamedPipeBuilder::new(addr)
-				.first(false)
-				.accept_remote(true)
-				.inbound(true)
-				.outbound(true)
-				.out_buffer_size(MAX_REQUEST_LEN)
-				.in_buffer_size(MAX_REQUEST_LEN)
-				.create()));
+		let mut connected_pipe = {
+			let mut pipe = self.waiting_pipe.lock().unwrap();
```

Comment:
The following also passes the test for me:
```rust
loop {
	if {
		let pipe = self.waiting_pipe.lock().unwrap();
		unsafe {
			! try!(pipe.connect_overlapped(overlapped.raw()))
		}
	} {
```
and then several lines below, 
```rust
let mut pipe = self.waiting_pipe.lock().unwrap();
std::mem::replace::<NamedPipe>(&mut *pipe,
```

Clearly mutexes are more short-lived and granular this way, but maybe it has some consequences  for Windows pipe handling somehow?